Added the CMake Install Command

Tyler-g-hudson opened the pull request:
CMake install command added, finishing off the major build steps.
Also changed the module name from `docker_cli` to `wigwam`.

Tyler-g-hudson committed:
docker_cli -> WIGWAM

Tyler-g-hudson committed:
Added `cmake-install` command

Tyler-g-hudson committed:
tiny fix

gmgunter commented on wigwam/_docker_cmake.py:
```diff
@@ -86,3 +86,43 @@ def cmake_build_dockerfile(base: str) -> str:
+
+def cmake_install_dockerfile(base: str, ld_lib: str) -> str:
+    """
+    Creates a dockerfile for installing with CMAKE.
```
```suggestion
    Creates a Dockerfile for installing with CMake.
```

gmgunter commented on wigwam/_docker_cmake.py:
```diff
@@ -86,3 +86,43 @@ def cmake_build_dockerfile(base: str) -> str:
+    ----------
+    base : str
+        The base image tag.
+    ld_lib : str
+        The name of the linking library file (e.g. "lib" or "lib64".)
```
Let's rename this parameter to `libdir`.

```suggestion
    libdir : str
        The base name of the directory where object libraries are stored
        (e.g. "lib" or "lib64"). This should match the value of `LIBDIR`
        in CMake's `GNUInstallDirs` module 
        (https://cmake.org/cmake/help/latest/module/GNUInstallDirs.html).
```

gmgunter commented on wigwam/_docker_cmake.py:
```diff
@@ -86,3 +86,43 @@ def cmake_build_dockerfile(base: str) -> str:
+    # Install the project and set the appropriate permissions at the target directory.
+    dockerfile += dedent(
+        f"""
+            USER root
```
Why are we running as `root` here? Is there an issue with installing isce3 as the default `MAMBA_USER`?

gmgunter commented on wigwam/_docker_cmake.py:
```diff
@@ -86,3 +86,43 @@ def cmake_build_dockerfile(base: str) -> str:
+        f"""
+            USER root
+            RUN cmake --build $BUILD_PREFIX --target install --parallel
+            RUN chmod -R 755 $BUILD_PREFIX
```
Why are we changing the permissions of the build prefix here?

Note that the package should be installed to `$INSTALL_PREFIX` -- not `$BUILD_PREFIX`. It's not totally clear to me why we should need to modify the permissions of either directory, though.

gmgunter commented on wigwam/_docker_cmake.py:
```diff
@@ -86,3 +86,43 @@ def cmake_build_dockerfile(base: str) -> str:
+            RUN chmod -R 755 $BUILD_PREFIX
+
+            USER $MAMBA_USER
+            RUN mkdir /tmp/Testing
```
Why are we creating this directory here? Doesn't seem like this belongs in the `cmake-install` step.

gmgunter commented on wigwam/_docker_cmake.py:
```diff
@@ -86,3 +86,43 @@ def cmake_build_dockerfile(base: str) -> str:
+            RUN mkdir /tmp/Testing
+
+            ENV LD_LIBRARY_PATH $LD_LIBRARY_PATH:$INSTALL_PREFIX/{ld_lib}
+            WORKDIR $BUILD_PREFIX
```
Why are we changing the working directory to the build prefix here?

Later on when we run the isce3 test suite, we will probably need to run from inside the build directory. But no need to change the working directory preemptively now.

gmgunter commented on wigwam/_docker_cmake.py:
```diff
@@ -86,3 +86,43 @@ def cmake_build_dockerfile(base: str) -> str:
+            USER $MAMBA_USER
+            RUN mkdir /tmp/Testing
+
+            ENV LD_LIBRARY_PATH $LD_LIBRARY_PATH:$INSTALL_PREFIX/{ld_lib}
```
In addition to setting the `LD_LIBRARY_PATH`, I think we also want to append the `packages` directory to the `PYTHONPATH` here

```suggestion
            ENV LD_LIBRARY_PATH $LD_LIBRARY_PATH:$INSTALL_PREFIX/{ld_lib}
            ENV PYTHONPATH $PYTHONPATH:$INSTALL_PREFIX/packages
```

gmgunter commented on wigwam/commands.py:
```diff
@@ -218,6 +228,46 @@ def compile_cmake(tag: str, base: str, no_cache: bool = False) -> Image:
+
+    image: Image = Image(prefixed_base_tag)
+
+    is_64_bit = test_image(image=image, expression='"$BUILD_PREFIX/lib64"')
```
> `is_64_bit`

Actually, the distinction between the name `lib` vs `lib64` doesn't have anything to do with whether Linux (or isce3) was 64-bit or not in this case. It's just due to a different naming convention among different distros.

I think macOS and Debian-based OSes both use "lib" but other Linux distros tend to use "lib64".

So let's rename this variable to something like `uses_lib64` to avoid confusion.

gmgunter commented on wigwam/commands.py:
```diff
@@ -218,6 +228,46 @@ def compile_cmake(tag: str, base: str, no_cache: bool = False) -> Image:
+    if is_64_bit:
+        lib = "lib64"
+    else:
+        lib = "lib"
```
Let's add an assertion here to double-check that we're doing the right thing.

```suggestion
        lib = "lib"
        assert test_image(image=image, expression='"$BUILD_PREFIX/lib"')
```

gmgunter reviewed (commented):
Needs a unit test. A good test would be to see if you can run `python -c "import isce3"` and `python -c "import nisar"` inside the generated image.

Tyler-g-hudson committed:
PR fixes

+ Added documentation for why to use USER root in cmake install
+ Added a check for the case where an image uses 'lib' and not 'lib64'
+ Added PYTHONPATH env variable change in cmake install
~ Changed some documentation in cmake install
- Remove workdir change in cmake install

Tyler-g-hudson committed:
Added unit test, fixed libdir bug

gmgunter reviewed (approved):
LGTM, thanks!
